Changed argparse to use sub-parser commands capture/restore, and gave…

michaeldbrandt opened the pull request:
… them a parent parser that they share the same arguments.  thdbbug is treated as a separate command with no parent.  Modified tests to account for slightly different calling convention, the command must come before the arguments for it.

michaeldbrandt committed:
Changed argparse to use sub-parser commands capture/restore, and gave them a parent parser that they share the same arguments.  thdbbug is treated as a separate command with no parent.  Modified tests to account for slightly different calling convention, the command must come before the arguments for it.

cameronwaterman commented on slmigrate/arghandler.py:
```diff
@@ -3,50 +3,43 @@
+    parent_parser.add_argument("--" + constants.migration_arg, "--directory", "--folder", help="Specify the directory used for migrated data", action="store", default=constants.migration_dir)
+    parent_parser.add_argument("--" + constants.source_db_arg, "--sourcedb", help="The name of the source directory when performing intra-databse migration", action="store", default=constants.source_db)
+
+    commands = parser.add_subparsers(dest='action')
```
Can we store the "action" attribute off as a const? It looks like it is being used as a raw string in multiple locations.

prestwick replied:
I agree with this.

cameronwaterman reviewed (commented):
Just commenting - I'll leave the approval up to Mark since it's his repo.

prestwick commented on slmigrate/arghandler.py:
```diff
@@ -3,50 +3,43 @@
-    parser.add_argument("--" + constants.source_db_arg, "--sourcedb", help="The name of the source directory when performing intra-databse migration", action="store", default=constants.source_db)
-    return parser
-
+def parse_arguments():
```
Why was `args` removed? This tool doesn't actually run with this removed. 
```
PS C:\Users\markblack\Documents\GitHub\systemlink-migration-sandbox> python .\systemlinkmigrate.py restore --test
Traceback (most recent call last):
  File ".\systemlinkmigrate.py", line 9, in <module>
    arguments = arghandler.parse_arguments(sys.argv[1:]).parse_args()
TypeError: parse_arguments() takes 0 positional arguments but 1 was given
```



Keep in mind that the test as structured now are unit tests not integration tests. Those need to be run by hand - basically just calling the tool like an end user.

prestwick commented on slmigrate/arghandler.py:
```diff
@@ -3,50 +3,43 @@
+  
+    commands.add_parser(constants.thdbbug.arg, help="Migrate tag history data to the correct MongoDB to resolve an issue introduced in SystemLink 2020R2 when using a remote Mongo instance. Use --sourcedb to specify a source database. admin is used if none is specfied",)
 
-def handle_unallowed_args(arguments):
```
I'm assuming this function was removed because sub parsers should do this for "free"? That's ok except for its still being called from systemlinkmigrate.py, which leads to an exception.

```
PS C:\Users\markblack\Documents\GitHub\systemlink-migration-sandbox> python .\systemlinkmigrate.py capture --test
Traceback (most recent call last):
  File ".\systemlinkmigrate.py", line 10, in <module>
    arghandler.handle_unallowed_args(arguments)
AttributeError: module 'slmigrate.arghandler' has no attribute 'handle_unallowed_args'

michaeldbrandt replied:
Sorry about that oversight.  I was so focused on getting the tests to pass that I forgot to update the slmigrate.py script.  I've done that now.

prestwick reviewed (changes requested):
@michaeldbrandt thank you for these updates and getting into the code. Please be sure that as you make changes you run some of the commands yourself when making changes. Tests are extremely important to prevent regressions and enable refactoring but nothing should prevent you from testing the final workflow as well.

michaeldbrandt committed:
Removed the hardcoding for the attr used for the arg parse subparser name in favor of a global constant.

michaeldbrandt committed:
Updated slmigrate.py to use the new argparser

michaeldbrandt committed:
Updated slmigrate.py to use the new argparser

michaeldbrandt committed:
Merge branch 'parent-arg-parser-commands' of https://github.com/prestwick/systemlink-migration-sandbox into parent-arg-parser-commands

michaeldbrandt committed:
Update python-app.yml

Changed branch name from 'master' to 'main' in order to get workflow to trigger correctly
